deepblast/metrics.py

```diff
@@ -0,0 +1,594 @@
+    p2 : np.array
+       Point cloud of M d-dimensional points
+    align_index : np.array
+       Translates i'th alingment point to the index in master
+
+    Returns
+    -------
+    Tuple of metrics
+
+    Notes
+    -----
+    The percentage of structural similarity (PSI) is defined as the
+    number of aligned amino acid pairs with Cα atoms that are closer in space
+    then 4 Å after optimal superposition normalized by the length of the
+    shorter chain in the alignment. The relevant PSI (rPSI) value does not
+    include fragments shorter than four aligned amino acid from the
+    calculated PSI value. The coordinate RMSD (cRMSD) is computed
+    for all aligned pairs after optimal superposition. The cRMSD (core) is
+    computed for those aligned pairs that contribute to the PSI value [35].
+    Here, PSI/rPSI provides a more detailed view of the alignment.
+    However, these values are length dependent.'''
+    if indicies is None:
+        indicies = range(np.shape(align_index)[1])
+    # get the shape
+    L_min = min( np.shape(master_p0)[0],np.shape(master_p1)[0])
+    L_aligned=np.shape(align_index)[1]
+    L_orientable = len(indicies)
+
+    TM_d0 = 1.24 * (L_min - 15) ** 0.333333 - 1.8  # for the TM score
+    TM_d02 = TM_d0 ** 2
+
+    p0 = master_p0[align_index[0]]
+    p1 = master_p1[align_index[1]]
+
+    p0aligned, p1aligned, G = kabsch_template_alignment(
+        p0, p1, p0[indicies], p1[indicies]
+    )
+    # rotation aligns just the selected atoms
+    # this is over the full alignment.
+    deviation2 = np.sum((p0aligned - p1aligned) ** 2, axis=1)
+    TM_score = np.sum(1.0 / (1.0 + deviation2 / TM_d02)) / L_min
+    # RMS over all
+    RMS = np.sqrt(np.sum(deviation2) / L_aligned)
+    # RMS over subset of resides contributing to orientation
+    oRMS = np.sqrt(np.sum(deviation2[indicies]) / L_orientable)
+
+    PSI_mask =  np.sqrt(deviation2) < (4.0 * UNIT)
+    L_PSI = np.sum(PSI_mask)
+    PSI = L_PSI / L_min
+
+    if L_PSI > 2: # RMS doesn't mean much otherwise
+        # RMS over resides contributing to PSI
+        # or should be re-align this?  could become tailchasing?
+        cRMS = np.sqrt(np.sum(deviation2[PSI_mask]) / L_PSI)
+    else:
+        cRMS = np.NaN  # fiction.
+    if seq0 is not None and seq1 is not None:
+        seq_aligned = np.array([ [seq0[i], seq1[j]] for i,j in align_index.T ])
+        aSeq_ident = np.sum( seq_aligned[:, 0] == seq_aligned[:, 1])
+        aSeq_ident /= L_aligned
+        oSeq_ident = np.sum(
+            seq_aligned[indicies, 0] == seq_aligned[indicies, 1]
+        )
+
+        oSeq_ident /= L_orientable
+
+        cSeq_ident = np.sum(
+            seq_aligned[PSI_mask, 0] == seq_aligned[PSI_mask, 1]
+        )
+
+        cSeq_ident /= L_PSI
+    else:
+        aSeq_ident = oSeq_ident = cSeq_ident = 0
+
+    c = 0 # run length tracker
+    aPSI = 0
+    # running sum of long runs with 4 or more residues in a row without gaps
+    for i in range( L_aligned):
```

Comment:
The offending lines start here. When I replace this with 
```suggestion
    for i in range( L_aligned - 1):
```
The out of index errors go away.